### ISSUE TITLE
Generating relative links to assets

### ISSUE THREAD

**Issue by aantron** [dmbaturin/soupault#24]
How does one generate correct links to assets, from pages that may be in multiple different subdirectories of a site? I believe that relative paths are necessary for this. The template

```html
<html>
<head>
  <link rel="stylesheet" href="style.css">
</head>
<body>
</body>
<html>
```

effectively expects a `style.css`in every subdirectory in which the template will be instantiated, so I think I need to replace the `style.css` based on the output path. I haven't (so far) found docs on doing this.

Using an absolute `/style.css` doesn't work:

- It doesn't work at all when viewing the site locally (unless, I suppose, I place `style.css` at the right place in the filesystem...)
- It won't work on a site deployed somewhere other than `/` under its domain.

**Comment by dmbaturin** [comment]
I made a site URL plugin: https://soupault.app/plugins/#site-url that allows inserting a prefix in every relative link (every `<a>`, `<img>`, and everything that can point to a URL).

**Comment by aantron** [comment]
AFAICT the plugin still won't solve the problem for local viewing of the generated site. It makes the site less portable instead.

By comparison, when Bisect_ppx generates its [coverage reports](https://aantron.github.io/bisect_ppx/demo/), it generates relative paths to static assets. For example, on [this page for `common.ml`](https://aantron.github.io/bisect_ppx/demo/common.ml.html), the path to the stylesheet is `coverage.css`, but on [this page for `lwt_unix/markup_lwt_unix.ml`](https://aantron.github.io/bisect_ppx/demo/lwt_unix/markup_lwt_unix.ml.html), the path is `../coverage.css`. This makes the coverage report completely portable, require no site configuration at all.

Something like this would be quite useful in Soupault.

Another drawback of the plugin is it requires being aware of Lua, which is extra mental load in case Soupault is being integrated in a project that already uses several different languages and formats.

I think this (first approximation, still not faimiliar enough) that this should be done by something like a built-in widget.

**Comment by dmbaturin** [comment]
Now I see what you mean. There seems to be some cultural bias against relative URLs. Most static site generators provide a built-in option to inject a site URL in links, but I've never seen one that would support making them relative. I suppose it's because most people are testing their sites by running a development web server in the build directory, whether SSG's built-in or something like `python -m http.server --directory build/` so links pointing at `/` are not an issue.

I do acknowledge your use case exists though: for coverage reports and similar, portability is a real issue since they can easily end up in any directory, unknown beforehand.

I'm not yet sure if I would want either this or site URL or both to be a built-in functionality of soupault. I'm not really against extending the built-in functionality, but my concern is that these things can be very hard to get _universally right_. Soupault is meant to be an antidote for SSGs like Hugo or Zola that have many things built-in but don't give you a way to change the way those things works.

Same goes for
>Another drawback of the plugin is it requires being aware of Lua, which is extra mental load

At least it's Lua rather than a Turing-complete abomination of a template processor. ;)

I'm not completely against ever making it a built-in though. I've written a plugin to experiment with it, from my quick test it does what you want, but needs more testing of course.

```
-- Adjusts relative links to point at the correct directory
-- as page depth changes.
-- e.g. href="main.css" will become href="../main.css" on "/about/index.html"
--
-- Has no special config options.
--
-- [widgets.make-relative-links]
--   widget = "site-url"
--
-- Author: Daniil Baturin
-- License: MIT

-- Plugin code

function make_parent_path(depth)
  local parent_path = ""
  while(depth > 0) do
    parent_path = parent_path .. "../"
    depth = depth - 1
  end

  return parent_path
end

function adjust_path(elem_name, attr)
  elems = HTML.select(page, elem_name)

  local index = 1
  while elems[index] do
    elem = elems[index]
    target = HTML.get_attribute(elem, attr)
    if target then
      -- Check if  a URL schema is present
      if not Regex.match(target, "^([a-zA-Z0-9]+):") then
        -- If not, check if it starts with a / or a .
        -- Links relative to the root and handmade relative links should be left untouched
        if (not Regex.match(target, "^/(.*)")) and (not Regex.match(target, "^\\.(.*)")) then
          -- If that file exists in the target dir, the link should remain relative
          if not Sys.file_exists(target_dir .. target) then
            -- Strip the build_dir prefix from the target dir
            local relative_target_dir = Regex.replace(target_dir, "^" .. build_dir, "")
            local dir_path = Regex.split(relative_target_dir, "/")

            local depth = size(dir_path)
            local parent_path = make_parent_path(depth)

            target = parent_path .. target
            HTML.set_attribute(elem, attr, target)
          end
        end
      end
    end
    index = index + 1
  end
end

adjust_path("a", "href")
adjust_path("link", "href")
adjust_path("img", "src")
adjust_path("script", "src")
adjust_path("audio", "src")
adjust_path("video", "src")
adjust_path("embed", "src")
adjust_path("object", "data")

```

**Comment by dmbaturin** [comment]
Of course, since someone who's unhappy with the built-in can always make their own plugin, perhaps there's no harm in making a built-in implementation.

The configuration options of that widget need some discussion though.

**Comment by aantron** [comment]
AFAIK the main reason people are running web servers during development when working with static sites is to have them translate requests for directories `foo` to `foo/index.html` for the pretty URLs. For simple purposes, however, it seems better if the site can be tested without a server (except of course links won't work well with pretty URLs &mdash; but at least the page itself can render with CSS and other assets). This is arguably **in** line with having Soupault tend towards lighter weight. No yucky web server to run :)

I'm working on a docs site at the moment. It would be very convenient to preview it trivially.

> There seems to be some cultural bias against relative URLs.

It's probably not against relative URLs, just for `index.html`. That makes me wonder if the relative URLs will still work from a location like `foo/`.

> built-in but don't give you a way to change the way those things work.

As I understand it, if it's a built-in *widget*, as I suggested (I didn't suggest just "built in" like Hugo, and I wouldn't want that :P), the user will have a way to change how it works &mdash; by simply not opting into it. You could generate it commented-out in the `--init` config file, for example.

For me, I think finding every `href`, `src`, etc., that starts with `/`, and replacing the path by a correct relative path, will work.

Probably does need some initial testing with pretty URLs locally vs. on server, though.

**Comment by aantron** [comment]
> not opting into it

...or whatever configuration settings the widget may have, of course.

**Comment by dmbaturin** [comment]
>It's probably not against relative URLs, just for index.html

Oh, I mean I don't see many `../` links in the wild, or anyone advocating them, even though they work perfectly in browsers.

Fair point about clean URLs. I'm using them on all my sites, but I also don't mind previewing with Python's web server since I have Python on all my machines anyway. 

The plugin I posted above works for both "clean" and "traditional" URLs, because soupault's `target_dir` variable in the plugin environment takes that setting into account—it's the dir where the generated page file will be saved.

Let me know when you have your site repo available so that I can build and test it.

Also, for the [iproute2 manual](https://baturin.org/docs/iproute2) I'm using a different approach with [inlining assets](https://github.com/dmbaturin/iproute2-cheatsheet/blob/master/plugins/inline-assets.lua), so that anyone can save the page locally without having to resort to converting it to MHT/CHM/whatever else. Maybe the two approaches can be combined.

**Comment by aantron** [comment]
As another data point, odoc generates relative URLs. For example, https://paurkedal.github.io/ocaml-caqti/caqti-lwt/Caqti_lwt/index.html points to `../../odoc.css`, and is (of course) hosted online. It can be viewed and worked on equally well locally without a web server. It is also accessible as

- https://paurkedal.github.io/ocaml-caqti/caqti-lwt/Caqti_lwt/
- https://paurkedal.github.io/ocaml-caqti/caqti-lwt/Caqti_lwt with a typical redirect by `*.github.io`

...and works on both.

**Comment by aantron** [comment]
I disagree with this (or misunderstand it):

```lua
-- Links relative to the root and handmade relative links should be left untouched
if (not Regex.match(target, "^/(.*)")) and (not Regex.match(target, "^\\.(.*)")) then
```

The links that begin with `/` are the ones that need to be rewritten, since they are the ones that make an assumption about the prefix of the path (that there isn't one). That assumption is broken by viewing from the filesystem, or by hosting e.g. on `me.github.io/myprojet`, where the prefix is `/myproject`, not empty.

Links containing URLs or relative paths aren't broken (except if Soupault otherwise rearranges things).

If Soupault (or any SSG) provides a way to rewrite all such links to relative links, it completely eliminates the need to make assumptions about the prefix or manipulate these assumptions with configuration.

**Comment by aantron** [comment]
The exception to this would be when authoring a site that is a sibling of some other resources under some other prefix, when you do need a clean absolute path to refer to them &mdash; but this could be handled at a coarse level by making the path rewriter opt-in, and maybe at a finer level, upon request, by adding some kind of filtering to the widget configuration, e.g. based on class or `data-` attribute.

Another way to implement this is to have Soupault substitute some variable to templates for the relative path to the site root, that could be used where needed.

For my use case, I prefer the widget bulk-rewriting method.

**Comment by aantron** [comment]
And sorry for the spam &mdash; I keep coming up with these things one at a time :) If you look at a typical odoc-generated docs site, you will see that *all* the links are relative, not just the CSS. For example, on the Caqti_lwt page I linked, look at the target of `future` or many of the other links.

**Comment by dmbaturin** [comment]
>I disagree with this (or misunderstand it)

That's one of my arguments in favor of making such things loadable plugins rather than built-in widgets. ;)

I was making an assumption that if a user writes `/main.css`, they specifically want to point to the server root, while if they write just `main.css`, they want it "relativized".

At the very least the widget should have a configurable option for that assumption. Perhaps a simple regex option will do.

**Comment by dmbaturin** [comment]
What do you think about this design?

```
[widgets.make-links-relative]
  widget = "relative-links"
  # Links whose URL matches this regex will be left untouched
  # The default is "anything that seems to have a URI schema
  exclude_url_regex = '^([a-zA-Z0-9]+):'
  
  # Before adjusting a link URL for the directory tree depth,
  # check if it needs to be done at all.
  # If a page has a link like <link rel="stylesheet" href="style.css">
  # and there's a file named style.css in the directory with the page,
  # the link likely doesn't need any adjustment.
  # However, people who know they don't have such custom assets
  # may want to disable those checks for performance and predictability.
  check_files = true

  # Asset files may not be stored in the site_dir with pages,
  # some people may create them with plugins or build scripts.
  # By default soupault will check for files in the page source dir,
  # but those people may prefer to check in the build dir instead.
  check_in_build_dir = false

  # Different HTML elements use different attributes for the URL
  # Some people may want to add custom elements 
  elements = {
    a = "href",
    link = "href",
    img = "src",
    ...
  }
```

**Comment by aantron** [comment]
Thanks!

> ```toml
>   # Links whose URL matches this regex will be left untouched
>   # The default is "anything that seems to have a URI schema
>   exclude_url_regex = '^([a-zA-Z0-9]+):'
> ```

As I undertand it, the widget wouldn't be able to make links with a URI scheme into relative links anyway. I see that this setting may have other uses, of course. I would set it to `^[^/]`, so exclude any links that don't begin with `/` &mdash; but then again, in my mind, the widget should exclude links that are already relative by default. So with my understanding of the issue (perhaps from a too-narrow point of view), the only links that would be eligible for rewriting in the  first place would be absolute paths (`/...`), Soupault would rewrite all of them by default, and this setting would allow excluding some of them.

> ```toml
>   check_files = true
>   check_in_build_dir = false
> ```

The links I want to rewrite (`/...`) always need adjustment for my purpose (portability), so I wouldn't use this setting. Again, with my understanding, I don't see why you would check &mdash; the only paths that wouldn't need rewriting are ones that are already relative, and it seems like it's best to leave them alone (at least by this widget; the user may indeed want to do other things to them like make them into full URLs).

The only exception to a `/...` path needing adjustment would be if there was some well-known resource outside the site but on the same domain I was trying to reach. This check wouldn't be able to discover such a link. Excluding it could be handled by `exclude_url_regex`.

I think the idea (as I interpret some of the design) to rewrite non-`/...` paths, especially with this check, will lead to  confusion. For example, what if there is a `main.html` template with a relative link to `style.css` entered by the user, instantiated in multiple directories &mdash; but both the site root and a subdirectory have a `style.css`. It seems like a strange combination of behaviors that in the root directory, the link will turn into `style.css`, in most subdirectories, it will become `../style.css`, but in one subdirectory, it will remain `style.css` and point to a different stylesheet. I think this relative link should just always be left as `style.css`.

Rewriting only absolute paths `/style.css` and doing no checks for file presence seems much more "statically predictable" somehow, as the presence of the `/`, if the user opts into this widget, is essentially a directive to make the link relative and point at the root `style.css`, no matter where the page instantiating the template is.

> ```toml
>   # Different HTML elements use different attributes for the URL
>   # Some people may want to add custom elements 
>   elements = {
>     a = "href",
>     link = "href",
>     img = "src",
>     ...
>   }
> ```

+1!

<br>

In summary, and IMO, I think the `check_files` options should be dropped, and I think the widget shouldn't touch non-`/...` links by default, and, of course, the whole widget should not be enabled by default. It should be possible to add various options about rewriting these other links with this widget later, upon someone's request &mdash; I wouldn't use them (at least now), and they worry me. They also seem somewhat more complicated in implementation than the other options (e.g. the need for `check_in_build_dir`, performance considerations), so I would suggest to defer implementing them until they are needed.

**Comment by aantron** [comment]
> I was making an assumption that if a user writes `/main.css`, they specifically want to point to the server root, while if they write just `main.css`, they want it "relativized".

I would like `/` to point at the site root (and this is the only option that makes sense for a portable site, especially one that may be viewed from the file system, as `/...` paths have no valid interpretation in such a site AFAIK). I do see that someone may want to write links pointing to the server root. I agree that would be well-handled by `exclude_url_regex`. If not, you could later (upon someone's request) add a toggle option to leave absolute paths completely alone, and interpret relative paths that don't begin with `./` as eligible for rewriting (or some other such scheme).

I think that someone using valid `/...` links already knows something about the structure of the overall server they are hosting the site on, so they will typically already have much less use for this widget &mdash; though they may still need it for some purpose. My main purpose, by contrast, is to make the site completely portable (at least in terms of locating assets; page links still won't work due to inability to serve `index.html` when accessing file systems). I think the defaults shouldn't harm that in any way, because AFAIK portability is the main use case of this widget (again, making allowance for the possibility that my interpretation is too narrow).

**Comment by aantron** [comment]
To give a bit more detail about my intended usage.

I don't want to manually write relative links, as they are fragile to both the origin and the target moving. I would prefer to write absolute paths (`/...`) only, and have the generator resolve them all into relative links.

I wonder if the cultural bias against relative links you were referring to is at play here &mdash; the issue is using relative links in source files. That's exactly what I would like to (mostly) avoid, by using a generator to interpret my absolute paths.

**Comment by dmbaturin** [comment]
>the only links that would be eligible for rewriting in the first place would be absolute paths (/...)

That depends on the site layout. From another point of view, _if_ the site is at the server root and soupault is used as a generator rather than a post-processor, then links that begin with a / will keep working at any depth, while links like `href="style.css"` that come from a template will need adjustment when soupault generates pages in subdirs from that template.

For the ultimate portability, one will want to rewrite all links that don't have a URI schema. Now that I think of it, when a link begins with a /, soupault should remove that / to make it usable at non-root web server level.

In other words, as Alan Perlis said, "one's constant is someone else's variable".

>It should be possible to add various options about rewriting these other links with this widget later, upon someone's request

In my experience, when people see an unconfigurable feature, they tend to assume it's opinionated rather than incomplete and that the way it works is what the developer thinks is the one true way it should work. Also from my experience, it's usually a correct interpretation: I had to fight tooth and nail to have authors of hardcoded behavior acknowledge that other use cases exist and rarely succeded, while adding a new option to already configurable code is rarely an issue.

Thus I believe it's better to acknowledge that there are different use cases right away.

Anyway, I'm going to look into the implementation this weekend and make a prototype.

**Comment by dmbaturin** [comment]
Took a bit longer than I hoped, but the prototype is in. Please test and tell me if you find any issues.

The element and attribute mapping is hardcoded for now.

Here's the config I used for testing it on my ownsite:

```
[widgets.relativize]
  widget = "relative_links"
  check_file = true # default is false
```

**Comment by dmbaturin** [comment]
Perhaps I should add something like `only_target_regex` for use cases like "relativize only links that start with a /" like you mentioned.

**Comment by aantron** [comment]
Thanks.



> Perhaps I should add something like `only_target_regex` for use cases like "relativize only links that start with a /" like you mentioned.

Is it not doing that now? That's what I would be testing, so I'll hold off for now.

**Comment by dmbaturin** [comment]
I've just added `only_target_regex` option. With `only_target_regex = "^/"` it will only adjust links that start with a slash.

The original implementation would relativize everything that doesn't match the default `exclude_target_regex = "^((([a-zA-Z0-9]+):)|#)"`, i.e. everything that doesn't have a URI schema and isn't a link to an anchor in the same page.

**Comment by aantron** [comment]
Thanks! It just tried it with `only_target_regex = "^/"` and it seems to work well.

**Comment by dmbaturin** [comment]
I've been planning to make an OPAM release just before you asked about relative links. Is there anything else you want to get working before a new release?

**Comment by aantron** [comment]
I am writing docs with links like `foo/index.html` so that they are portable between filesystems and the web. I was considering making a postprocessor that can find and drop trailing `index.html` on relative links automatically, and which can be enabled when deploying to the web. However, I haven't reached the stage where I am testing that yet, so I don't have good suggestions for it, and I haven't confirmed the need.

**Comment by dmbaturin** [comment]
>and which can be enabled when deploying to the web

You can do different things in different scenarios using "build profiles". You can limit a widget to specific build profile and then set the profile from the command line. See https://soupault.app/reference-manual/#build-profiles

If you have no more immediate issues, I'll make some tweaks to the code and make a release. If you are happy with the way this widget works now, please close the issue.

**Comment by aantron** [comment]
Thank you!

**Comment by dmbaturin** [comment]
I've reworked the code a bit to make it OS-agnostic (i.e. work with Windows paths as well as normal paths). Could you re-test to make sure I didn't break anything?

**Comment by aantron** [comment]
Seems to still work.
